### DISCUSSION TITLE
Indentation preservation interpolated string values

### DISCUSSION THREAD

**Discussion by mitchdenny** [dotnet/csharplang#7007] [Language Ideas]
When writing a code generator, it is often convenient to use the new multi-line raw strings feature. Multi-line raw strings already preserve whitespace based on the indentation level of the trailing closing string literal quotes (```"""```);

However, consider the following block of code:

```csharp
var innerBlock = """
if (int.TryParse(input, null, out var input))
{
    return input * input;
}
else 
{
    throw Exception();
}
""";

var outerBlock = $$"""
public int Square(string inputString)
{
    {{innerBlock}}
}

Console.WriteLine(outerBlock);
""";
```

The output this would produce is:

```
public int Square(string inputString)
{
    if (int.TryParse(input, null, out var input))
{
    return input * input;
}
else
{
    throw Exception();
}
}
```

Basically, the first line of the inner block is where we would expect, but the subsequent lines are not where you want. Ideally the output would look like this:

```
public int Square(string inputString)
{
    if (int.TryParse(input, null, out var input))
    {
        return input * input;
    }
    else
    {
        throw Exception();
    }
}
```

I don't know what the solution might look like here, but I think that if string interpolation logic could inspect the type of the interpolated string placeholder it could kick in a specific behavior that allowed for indentation preservation. Perhaps something like:

```csharp
var innerBlock = (IInterpolationContext context) => context.AppendLines("""
if (int.TryParse(input, null, out var input))
{
    return input * input;
}
else 
{
    throw Exception();
}
""", Indentation.Preserve);

var outerBlock = $$"""
public int Square(string inputString)
{
    {{innerBlock}}
}
""";

Console.WriteLine(outerBlock);
""";
```

I'm not too wedded to the exact mechanism, but more the scope for simplification of code-gen code this might provide. One idea around the ```IInterpolationContext``` type is that implementors could create their own implementations, and based on the signature of the ```Func<...>``` the compiler would insert code to instansiate the ```IInterpolationContext``` implementor. This would allow the context to provide a range of APIs which are purpose designed for the code-gen scenario but still allow calling down into the base infrastructure.

**Comment by HaloFour** [comment]
This feels like a tooling or library concern to me.  For example, .NET already provides `System.CodeDom.Compiler.IndentedTextWriter` for prepending each line with a configured indentation depth.  T4 had similar with `PushIndent`/`PopIndent`.  There are numerous examples online about how you could also roll your own: [A Drop-in Indented TextWriter in C#](https://www.codeproject.com/Tips/5315915/A-Drop-in-Indented-TextWriter-in-Csharp)

**Reply by mitchdenny** [reply]
That same argument could be made for the multi-line literals that have recently been added to the language.

What I'm proposing is useful for the same reason multi-line literals are useful in that it allows you the ability to see the whole block of text you are emitting formatted as you would write it by hand - it makes it easy to reason over. The only difference between the current capabilities and what I'm proposing is that it provides a mechanism to potentially preserve indentation.

I can see what is proposed here being useful for writing tooling for sure.

**Reply by CyrusNajmabadi** [reply]
> That same argument could be made for the multi-line literals that have recently been added to the language.

A core reason for that was to not need to escape, and to allow indented literals *without* emitted indentation.  Neither of those hold here.  You can have a multi-line, indented, interpolated, raw-string literal, and you don't need escaping with it.  The question of what to do with 'holes' though is a separate concern.

As these are *interpolated* strings, you could actually solve this *today*, without any language support, just using an [interpolated string handler](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/interpolated-string-handler) that took the values for the holes and properly indented them as it processed them.

**Comment by jnm2** [comment]
Adjusting indentation at runtime like this is also not as efficient as this code, which I have tended to prefer anyway when considering the blocks as a whole:

```cs
var innerBlock = """
    if (int.TryParse(input, null, out var input))
    {
        return input * input;
    }
    else 
    {
        throw Exception();
    }
""";

var outerBlock = $$"""
public int Square(string inputString)
{
{{innerBlock}}
}
""";
```
